### ISSUE TITLE
no "learning" effect: loss calculation converges, but stagnation in reward 

### ISSUE THREAD

**Issue by akopf82** [werner-duvaud/muzero-general#119]
I've performed several tests (different games) using the initial configuration, and i always get mean values (Total reward/mean values [plot 2 / tictactoe]) alternating around 0. So there seems to be no "learning" effect - this is also what i get when i start games against the trained network... more or less randomized moves.
<img width="783" alt="tmp" src="https://user-images.githubusercontent.com/39181776/104812572-6eae6080-5803-11eb-8147-15d29ce424ac.png">
The loss calculation looks realistic to me (also compared to your screenshots), but the reward doesn't show progress over 10k or even 100k iterations. 
This observation also applies to the following modifications:
- mutliple GPU / one GPU
- varying amout of num_workers
- reduzed self.support_size
- "resnet" / "fullyconnected"
- "Adam" or "SGD"
- self.root_exploration_fraction 0.25 / 0.3

System: Python 3.7.6 , CentOS, Xeon(R) Gold + 2 x V100 -> no hardware limitation.

Do you have suggestions on how to get to the root of the problem?

**Comment by mokemokechicken** [comment]
Hi,

After much trial and error, I was able to get good results with tictactoe and connect4.

I haven't fully tested it yet, but I've created a branch with a few patches. Please give it a try.
https://github.com/mokemokechicken/muzero-general/tree/all_patch_for_master

### TicTacToe 
[Detail](https://github.com/mokemokechicken/muzero-general/tree/master/good_results/tictactoe-2021-01-18--01-16-32)
 ![image](https://user-images.githubusercontent.com/606549/104933054-c1e30780-59eb-11eb-94ca-1f627323628e.png)

### Connect4
[Detail](https://github.com/mokemokechicken/muzero-general/tree/master/good_results/connect4-2021-01-18--07-07-21)
![image](https://user-images.githubusercontent.com/606549/104933375-2c944300-59ec-11eb-891e-088a01679fd5.png)

**Comment by akopf82** [comment]
@mokemokechicken Thanks for this tests and the parameterlist - i can reproduce your results (tested on connect4). 
However, i discovered, that a pure adaption of the parameter based on the current [master branch](https://github.com/werner-duvaud/muzero-general) won't work! According to a fast vimdiff there seem to be at least adaptions in the temperature calculation:
<img width="824" alt="tmp" src="https://user-images.githubusercontent.com/39181776/105169012-35cffd80-5b1b-11eb-9ffc-dd1ef8b55fcf.png">
Are there further deviations compared to the current master branch?  The next few days i'll run some tests with parameter suggestions provided by [Julian](http://www.furidamu.org/blog/2020/12/22/muzero-intuition/).

**Comment by mokemokechicken** [comment]
@akopf82 
I'm glad you were able to reproduce it!
I am now checking which patch is required one by one.
The current thinking is,

- [PR-108](https://github.com/werner-duvaud/muzero-general/pull/108) is necessary.
- [this MCTS change](https://github.com/werner-duvaud/muzero-general/compare/master...mokemokechicken:all_patch_for_master#diff-61230cd65f2844876bcefd9377561a3d7f9ee1e96a3d2809e827b616c87a0848) is also necessary to improve the accuracy.

The other changes (temperature, etc.) were just implemented on the spur of the moment, and may not have much effect.

Increasing the speed of self-play while increasing num_simulation may be the best way to go.

**Comment by akopf82** [comment]
So these are my reward statistics based on your current branch. 
<img width="623" alt="tmp" src="https://user-images.githubusercontent.com/39181776/105226118-db0cc500-5b5f-11eb-8a7d-e719530c4ef4.png">
This looks really impressive, however the player strength is very weak:  
 Root value for player 0: -0.01
 Played action: Play column 4
    o
    o
    o o
    x x o   x x
 x o o o   x x
 o x o o   x x
 --------------
 1 2 3 4 5 6 7
This was one result of the self-play mode ... so a lot of chances to win have been missed... have you performed similar tests?

**Comment by mokemokechicken** [comment]
My self-play movie: https://youtu.be/fMla3FJCybY   (num_simlation is set to 250 to make it look better 😄 )
I think this is a good game.
Even if num_simulation is 50, it is not so weak.
It used to miss such decisive moves often, but with recent models, I rarely see them.

※ Is the Model loaded correctly?

By the way, I just found(and fix) a big bug in my master branch for self-play, but I don't think it affected the result of that mode because it doesn't affect if temperature is 0.

**Comment by akopf82** [comment]
※ Is the Model loaded correctly?
Honestly, i'm not sure - i use the standard replay functionality (4->connect4, 4->Play against MuZero), but there is no choice of the trained model, isn't it?

**Comment by mokemokechicken** [comment]
Oh, you missed an important step.

after "4->connect4", 
select "1. Load pretrained model" -> choice a dir number in `results/` or specify the path to the model,
and then "4->Play against MuZero" are correct steps.

**Comment by mokemokechicken** [comment]
I sent PRs for good training in two player game.
I think these are important.

- [PR-108](https://github.com/werner-duvaud/muzero-general/pull/108)
- [PR-121](https://github.com/werner-duvaud/muzero-general/pull/121)
- [PR-122](https://github.com/werner-duvaud/muzero-general/pull/122)

**Comment by akopf82** [comment]
Oh yes, it works :) thanks for the hint to use first "Load pretrained model"

**Comment by akopf82** [comment]
Based on following modifications (basis of [mokemokechicken](https://github.com/mokemokechicken/muzero-general/tree/master/good_results/connect4-2021-01-18--07-07-21)), i reach a "convergence" already after 30k iterations:

self.root_dirichlet_alpha = 0.2 -> 1   
self.num_unroll_steps = 5 -> 15  
self.PER_alpha = 0.5 -> 0.95 

<img width="415" alt="tmp" src="https://user-images.githubusercontent.com/39181776/105504339-eed33b00-5cc7-11eb-838f-297903dc0e76.png">

**Comment by alexZajac** [comment]
Hi @mokemokechicken ! 
I followed closely this issue as I am getting unexpected progress on the game of Gomoku, using the [current master branch.](https://github.com/werner-duvaud/muzero-general/tree/master) (trained on Google Colab).

Here are the tensorboard results:

![rewards](https://user-images.githubusercontent.com/38042033/108272871-1c01f480-7173-11eb-9bf7-7f88a0dd7332.PNG)
![games](https://user-images.githubusercontent.com/38042033/108272887-22906c00-7173-11eb-8f8f-e551e7918d8c.PNG)
![loss](https://user-images.githubusercontent.com/38042033/108272894-245a2f80-7173-11eb-905f-dddb658c9cf4.PNG)

I know that the Gomoku game was recently tweaked and I wondered if you think applying a similar set of parameters to the ones you had with connect4 would help?

Thank you!
